src/Moq/Setup.cs

```diff
@@ -169,11 +179,35 @@ protected virtual void ResetCore()
 		{
 		}
 
+		public void Verify(bool recursive = true)
+		{
+			this.Verify(recursive, setup => !setup.IsOverridden && !setup.IsConditional && setup.IsVerifiable);
+		}
+
+		public void VerifyAll()
+		{
+			this.Verify(recursive: true, setup => !setup.IsOverridden && !setup.IsConditional);
+		}
+
+		private void Verify(bool recursive, Func<ISetup, bool> predicate)
+		{
+			foreach (Invocation invocation in this.mock.MutableInvocations)
+			{
+				invocation.MarkAsVerifiedIfMatchedBy(setup => setup == this);
+			}
```

Comment:
Should this only happen if verification actually succeeds? Same question goes for the corresponding loop in `Mock.Verify`.

Reply:
While the current logic isn't perfect, it's generally good enough: if verification fails, an exception gets thrown anyway; it's unlikely someone catches that exception (Moq documentation explicitly recommends not doing that!) and then goes on to inspect the mock's state.

There may possibly be some odd corner cases regarding the verification status of invocations on a setup's inner mock; let's deal with those when they become an actual problem.